app/ide-desktop/lib/client/src/authentication.ts

```diff
@@ -0,0 +1,140 @@
+ * - The application emits an Electron {@link OPEN_URL_EVENT}.
+ * - The {@link OPEN_URL_EVENT} listener checks if the {@link URL} is a deep link.
+ * - If the {@link URL} is a deep link, the {@link OPEN_URL_EVENT} listener prevents Electron from
+ * handling the event.
+ * - The {@link OPEN_URL_EVENT} listener then emits an {@link ipc.channel.openDeepLink} event.
+ * - The {@link ipc.channel.openDeepLink} listener registered by the dashboard receives the event.
+ * Then it parses the {@link URL} from the event's {@link URL} argument. Then it uses the
+ * {@link URL} to redirect the user to the dashboard, to the page specified in the {@link URL}'s
+ * `pathname`. */
+import * as electron from 'electron'
+import * as ipc from 'ipc'
+import * as shared from '../shared'
+import opener from 'opener'
+
+
+
+// =================
+// === Constants ===
+// =================
+
+/** Name of the Electron event that is emitted when a URL is opened in Electron (e.g., when the user
+ * clicks a link in the dashboard). */
+const OPEN_URL_EVENT = 'open-url'
+
+
+
+// ========================================
+// === Initialize Authentication Module ===
+// ========================================
+
+/** Configures all the functionality that must be set up in the Electron app to support
+ * authentication-related flows. Must be called in the Electron app `whenReady` event. */
+export const initAuthenticationModule = (
+    window: () => electron.BrowserWindow | null,
+) => {
+    initIpc()
+    initOpenUrlListener(window)
+}
+
+/** Registers an Inter-Process Communication (IPC) channel between the Electron application and the
+ * served website.
+ *
+ * This channel listens for {@link ipc.channel.openUrlInSystemBrowser} events. When this kind of
+ * event is fired, this listener will assume that the first and only argument of the event is a URL.
+ * This listener will then attempt to open the URL in a cross-platform way. The intent is to open
+ * the URL in the system browser.
+ *
+ * This functionality is necessary because we don't want to run the OAuth flow in the app. Users
+ * don't trust Electron apps to handle their credentials. */
+const initIpc = () => {
+    electron.ipcMain.on(ipc.channel.openUrlInSystemBrowser, (_event, url) => opener(url))
+}
+
+/** Registers a listener that fires a callback for `open-url` events, when the URL is a deep link.
+ *
+ * This listener is used to open a page in *this* application window, when the user is
+ * redirected to a URL with a protocol supported by this application.
+ *
+ * All URLs that aren't deep links (i.e., URLs that don't use the {@link shared.DEEP_LINK_SCHEME}
+ * protocol) will be ignored by this handler. */
+const initOpenUrlListener = (
+    window: () => electron.BrowserWindow | null,
+) => {
+    electron.app.on(OPEN_URL_EVENT, (event, url) => {
+        const parsedUrl = new URL(url)
+
+        if (parsedUrl.protocol !== `${shared.DEEP_LINK_SCHEME}:`) {
+            return
+        }
+
```

Comment:
why these newlines?

Reply:
Again, to separate the `event.preventDefault()`, which is more of a side-effect than the intended logic.

Reply:
In such cases in our Rust codebase, we are just refactoring the body of the function to a few, well named sub-functions, so reading them line-by-line explains everything there. We can leave it this way, but this is an exception and we are not really using such newlines in other parts of the codebase.